src/librustc_data_structures/indexed_vec.rs

```diff
@@ -377,9 +373,21 @@ macro_rules! newtype_index {
             @vis          [$v]
             @debug_format [$debug_format]
                           $($tokens)*);
-        impl Decodable for $type {
-            fn decode<D: Decoder>(d: &mut D) -> Result<Self, D::Error> {
-                d.read_u32().map(Self::from)
+        newtype_index!(@decodable $type);
+    );
+
+    (@decodable $type:ident) => (
+        impl $type {
+            fn __decodable__impl__hack() {
+                mod __more_hacks_because__self_doesnt_work_in_functions {
```

Comment:
Heh, feel free to add a comment about using `const _ =` instead or something.
But also, why not just use absolute paths?
E.g. `impl ::serialize::Decodable for super::$type {` and `::serialize::Decoder::read_u32(d)`.

Reply:
If you use `decl_macro` you could even rely on hygiene, but idk how @nikomatsakis and @petrochenkov feel about using that in the compiler.

Reply:
Well... I wouldn't bank on `decl_macro` staying the same given that it has barely gone through an experimental RFC...

Reply:
absolute paths don't work, because not all crates have `extern crate serialize;`. Some have `extern crate serialize as rustc_serialize;`

Reply:
Right, I was hoping that wouldn't be a problem but... we'll have to wait for serde/some rustc-specific custom serialization solution, post-#49219.